### DISCUSSION TITLE
Collision Penetration issues

### DISCUSSION THREAD

**Discussion by TalhaEjaz03** [bepu/bepuphysics1#15] [Q&A]
Hi Norbo,
So we have run into a little problem that we are not able to solve yet.
The game objects when they collide with each other they penetrate inside
and also when using continuous detection mode when objects collide they sometimes jitter and move erratically. This happens more frequently when objects have increased momentum. 
Is there a way to resolve this issue and make the collisions more rigid. 
Can we remove the Softness on collision events??
Kindly guide us
Regards

`using System.Collections;
using BEPUphysics.CollisionTests.CollisionAlgorithms;
using BEPUphysics.CollisionTests.CollisionAlgorithms.GJK;
using BEPUphysics.Constraints;
using BEPUphysics.Materials;
using BEPUphysics.Settings;
using UnityEngine;
using PSpace = BEPUphysics.Space;
using PVector3 = BEPUutilities.Vector3;
using FixMath.NET;
using System.Collections.Generic;
using BEPUphysics.Entities;
using BEPUphysics.NarrowPhaseSystems;
public class PhysicsManager : MonoBehaviour
{

    public static PhysicsManager Instance { get; private set; }
    public Vector3 gravity = new Vector3(0f, -9.81f, 0f);
    public int Bepu_DefaultMinimumIterationCount = 1;
    public int Bepu_IterationLimit = 10;
    public bool Bepu_UseSimplexCaching = false;
    public bool Bepu_UseExtraExpansionForContinuousBoundingBoxes = false;
    public float Bepu_VelocityLowerLimit = 0.5f;
    public float Bepu_MaximumPenetrationRecoverySpeed = 2.0f;
    public float Bepu_BouncinessVelocityThreshold = 2.0f;
    public float Bepu_StaticFrictionVelocityThreshold = 0.2f;
    public float Bepu_ContactInvalidationLength = 0.1f;
    public float Bepu_ContactMinimumSeparationDistance = 0.03f;
    public float Bepu_MaximumContactDistance = 0.1f;
    public float Bepu_DefaultMargin = 0.04f;
    public float Bepu_AllowedPenetration = 0.01f;
    public float Bepu_DefaultMinimumImpulse = 0.001f;
    public int Bepu_maxTimeStep = 3;
    public float UpdateStepValue;
    public List<Entity> AllEntities = new List<Entity>();

    public PSpace Space { get; set; }

    bool GameStart = false;
    private void Awake()
    {
        GameStart = false;
        if (Instance == null)
        {
            Instance = this;
        }
        else if (Instance != this)
        {
            //Debug.Log("Instance Destroyed");
            Destroy(this.gameObject);
        }

        Init();

    }
    private void Start()
    {
        StartCoroutine(gameStarter());
    }
    private void Update()
    {
        //Space.Update((Fix64)0.1f);
        //Space.Update();

        if (GameStart == false)
        {
            Space.Update(0);
            //Debug.Log("dsadsadsadasd");

        }
        else if (Time.timeScale != 0)
        {
            //Space.Update((Fix64)(Time.fixedDeltaTime));
            Space.Update();
            //Space.Update((Fix64)UpdateStepValue);
        }

    }

    IEnumerator gameStarter()
    {

        Time.timeScale = 0;
        yield return new WaitForSecondsRealtime(3f);
        GameStart = true;
        Time.timeScale = 1;
        StopAllCoroutines();
        yield return null;
    }

    private void Init()
    {

        Space = new PSpace();
        Space.ForceUpdater.Gravity = new PVector3((Fix64)gravity.x, (Fix64)gravity.y, (Fix64)gravity.z);
        MaterialManager.MaterialBlender = delegate (BEPUphysics.Materials.Material a,
            BEPUphysics.Materials.Material b, out InteractionProperties properties)
        {
            properties.KineticFriction = BEPUutilities.MathHelper.Max(a.KineticFriction, b.KineticFriction);
            properties.StaticFriction = BEPUutilities.MathHelper.Max(a.StaticFriction, b.StaticFriction);
            properties.Bounciness = BEPUutilities.MathHelper.Min(a.Bounciness, b.Bounciness);
        };
        ApplySettings();
    }

    private void ApplySettings()
    {

        SolverSettings.DefaultMinimumIterationCount = Bepu_DefaultMinimumIterationCount;// by def 1
        Space.Solver.IterationLimit = Bepu_IterationLimit;    // by def 10
        Space.TimeStepSettings.MaximumTimeStepsPerFrame = Bepu_maxTimeStep;
        Space.TimeStepSettings.TimeStepDuration = (Fix64)0.03f;
        GeneralConvexPairTester.UseSimplexCaching = Bepu_UseSimplexCaching;    // by def False
        MotionSettings.UseExtraExpansionForContinuousBoundingBoxes = Bepu_UseExtraExpansionForContinuousBoundingBoxes;   // by def False

        // Set all the scaling settings back to their defaults.
        Space.DeactivationManager.VelocityLowerLimit = (Fix64)Bepu_VelocityLowerLimit;  // by Def 0.5f
        //Space.DeactivationManager.MaximumDeactivationAttemptsPerFrame = 1000;
        NarrowPhaseHelper.Factories.BoxSphere.EnsureCount(100);
        NarrowPhaseHelper.Factories.SphereSphere.EnsureCount(100);//16384
        CollisionResponseSettings.Softness = (Fix64)0.001;  // For softness of objects on collisionS
        CollisionResponseSettings.MaximumPenetrationRecoverySpeed = (Fix64)Bepu_MaximumPenetrationRecoverySpeed;   // by def 2
        CollisionResponseSettings.BouncinessVelocityThreshold = (Fix64)Bepu_BouncinessVelocityThreshold;   // by def 1
        CollisionResponseSettings.StaticFrictionVelocityThreshold = (Fix64)Bepu_StaticFrictionVelocityThreshold;    // by def 0.2f
        CollisionDetectionSettings.ContactInvalidationLength = (Fix64)Bepu_ContactInvalidationLength;  // by def 0.1f
        CollisionDetectionSettings.ContactMinimumSeparationDistance = (Fix64)Bepu_ContactMinimumSeparationDistance;    // by def 0.03f
        CollisionDetectionSettings.MaximumContactDistance = (Fix64)Bepu_MaximumContactDistance;    // by def 0.1f
        CollisionDetectionSettings.DefaultMargin = (Fix64)Bepu_DefaultMargin;   //by def 0.04f
        CollisionDetectionSettings.AllowedPenetration = (Fix64)Bepu_AllowedPenetration;    // by def 0.01f
        SolverSettings.DefaultMinimumImpulse = (Fix64)Bepu_DefaultMinimumImpulse;    // by def 0.001f
        // Space epsilons back to defaults

        // Leave below values as they are
        BEPUutilities.Toolbox.Epsilon = (Fix64)1e-7f;
        BEPUutilities.Toolbox.BigEpsilon = (Fix64)1e-5f;
        MPRToolbox.DepthRefinementEpsilon = (Fix64)1e-4f;
        MPRToolbox.RayCastSurfaceEpsilon = (Fix64)1e-9f;
        MPRToolbox.SurfaceEpsilon = (Fix64)1e-7f;
        PairSimplex.DistanceConvergenceEpsilon = (Fix64)1e-7f;
        PairSimplex.ProgressionEpsilon = (Fix64)1e-8f;

    }

}
`

**Comment by RossNordby** [comment]
>The game objects when they collide with each other they penetrate inside
and also when using continuous detection mode when objects collide they sometimes jitter and move erratically. This happens more frequently when objects have increased momentum.

I'd probably have to see a reproduction of the behavior in the demos to precisely diagnose it, but if I had to guess, this is related to a long timestep compared to motion. If an object can move a significant fraction of its entire size in a single timestep, it will tend to integrate into collision. v1's CCD can mitigate this some, but not entirely, especially in pathological cases. Further, if v1's CCD is pushed to its limits, it will tend to exhibit visible motion clamping artifacts (jerkiness). V2 doesn't have these issues, but alas.

If the above is correct, your best bet may be either to limit velocities or take more and shorter timesteps. That will cost more, but it will definitely improve behavior.

>Can we remove the Softness on collision events??

You can reduce the `CollisionResponseSettings.Softness`, but I'm not sure if that's what you actually need. Contact constraint softness is primarily relevant when there are heavy continuous forces smooshing stuff together. Reducing softness can lead to instability and jitter in difficult to solve situations.

**Reply by TalhaEjaz03** [reply]
Thanks man 
We solved the jerks and jitter issues 
that was a great tip
